returnn/tf/layers/basic.py

```diff
@@ -1645,6 +1645,113 @@ def get_out_data_from_opts(cls, name, sources, start=None, window_start=None, wi
     return out
 
 
+class RandIntLayer(LayerBase):
+  """
+  Generates random numbers using ``tf.random.uniform``
+  """
+  layer_class = "rand_int"
+
+  # noinspection PyUnusedLocal
+  def __init__(self, shape, maxval, minval=0, dtype="int32", seed=None, **kwargs):
+    """
+    :param tuple[DimensionTag|int]|list[DimensionTag|int] shape: desired shape of output tensor
+    :param int maxval: upper bound (exclusive) on range of random values
+    :param int minval: lower bound (inclusive) on range of random values
+    :param str dtype: type of the output. For random ints, int32 and int64 make sense, but could also be floats
+    :param int|None seed: random seed
+    """
+    from returnn.tf.util.data import DimensionTag
+    super(RandIntLayer, self).__init__(**kwargs)
+
+    seed = seed if seed is not None else self.network.random.randint(2 ** 31)
+
+    shape_parsed, _ = self._parse_shape(shape)
+    shape_parsed = tuple(s if not isinstance(s, DimensionTag) else self.get_batch_dim() for s in shape_parsed)
```

Comment:
This is very unintuitive.
You could maybe pass `batch_dim` to the `_parse_shape` function. But don't do such post edit here.